### PR TITLE
[installer]: move the kots install script into a bash file in Installer

### LINKED ISSUE #12197: move the bash script from installer-job to Installer container image
The bash script in the KOTS installer job has become far bigger than was ever intended. Having this written in the `args` section of a Kubernetes job is now working against us and we lose code completion, shellcheck by having it in there.

Move the script in the job into `/scripts` directory in the Installer and copy these into the container image at build stage so these can be run.

The `{{repl }}` options that are rewritten by KOTS at runtime will need to be passed to the script as environment variables, ensuring that no sensitive values are passed in this way. All efforts should be taken to remove the need for these in the scripts entirely and use secrets for these if they cannot be avoided entirely.

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Description

Moves the bash script in the installer job into the installer image. This is done to make the application more reliable, testable and readable. This makes no fundamental changes to how the job works except for fixing ShellCheck errors and changing all the `{{repl }}` references to environment variables.

This also improves upon the wording and gives troubleshooting steps in an effort to reduce the (small, but not insignificant) number of "help, my installation-status pod is in an errored state" issues we see raised.

## Related Issue(s)

Fixes #12197
Fixes https://github.com/gitpod-io/gitpod/issues/12259

## How to test

Install via KOTS.

I've checked this against all permutations that I can work out (k3s, AKS, EKS, GKE) and works on all.

## Release Notes

```release-note
[installer]: move the kots install script into a bash file in Installer
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

File: install/kots/manifests/gitpod-kots-config.yaml
Comment (corneliusludmann) [inline install/kots/manifests/gitpod-kots-config.yaml:13]: :bulb: (idea; non-blocking) Asking Replicated for a generic feature for this

---

Like this approach a lot.

I wonder if we should ask Replicated whether there is a built-in way to get the whole KOTS config as config map or if not, create a feature request. A template function that returns the whole KOTS config as YAML would probably be good enough.

I think, that creating a generic solution in KOTS should not be that hard for Replicated. It would make our lives much easier to maintain this.

What do you think?

_However, that does not block this pull request at all!_

Comment (MrSimonEmms) [reply install/kots/manifests/gitpod-kots-config.yaml:13]: It is possible to get this as a configmap, which we indeed do in the [support bundle](https://github.com/gitpod-io/gitpod/blob/main/install/kots/manifests/kots-support-bundle.yaml#L109,L122).

The problem with this as an approach is the KOTS config will have many sensitive pieces of data (passwords for DB/Registry/Storage etc) which we don't actually need for the Installer, as these are sent into secrets that the Installer then references.

We could of course store this as a secret rather than a configmap, but this would still contain information that we don't actually need for the Installer. I do see the benefit to not having to manage this though, but this then has the danger of thinking we're exposing something that we're not and errors being caused.

I'd propose that, for this PR, we keep it as-is and continue to discuss this subject when doing the general refactoring that we're doing.

Comment (corneliusludmann) [reply install/kots/manifests/gitpod-kots-config.yaml:13]: > I'd propose that, for this PR, we keep it as-is and continue to discuss this subject when doing the general refactoring that we're doing.

100 % agree.

**Comment by MrSimonEmms** [comment]
This is ready for test/review. I will leave the "hold" status as think this would benefit from multiple testers/approvers

File: install/kots/manifests/gitpod-kots-config.yaml
Comment (Pothulapati) [inline install/kots/manifests/gitpod-kots-config.yaml:12]: This looks good overall on a macro level, but my main concern is that these options are now publicly visible right? esps the registry, database values being here?

Can we use a secret instead, and then load the env in the `gitpod-installer-job` using a `secretReg`? :thinking: WDYT?

Comment (Pothulapati) [reply install/kots/manifests/gitpod-kots-config.yaml:12]: AHHHH, Right! They aren't changed. They follow the existing method of loading through secrets! 👍🏼

Comment (MrSimonEmms) [reply install/kots/manifests/gitpod-kots-config.yaml:12]: Yeah, I've tried to avoid changing anything significantly in this PR - that comes later :)

File: install/installer/scripts/kots-install.sh
Comment (Pothulapati) [inline install/installer/scripts/kots-install.sh:74]: *Definitely not asking in this PR*, but a next iteration of this change would be to move all the config settings into a separate script so that we can validate the end config as a pre-flight check right? Should be simple enough and hence solving https://github.com/gitpod-io/gitpod/issues/12015 :thinking:

File: install/installer/scripts/kots-install.sh
Comment (Pothulapati) [inline install/installer/scripts/kots-install.sh:118]: nit: Should it be called `REG_CUSTOM_DOCKER_CONFIG_ENABLED`? Would prefer to have `CONFIG` to be more specific.

File: install/installer/scripts/kots-install.sh
Comment (Pothulapati) [inline install/installer/scripts/kots-install.sh:140]: :thumsup: Like that this is now routed through a secret, instead of building it here!

File: install/installer/scripts/kots-install.sh
Comment (Pothulapati) [inline install/installer/scripts/kots-install.sh:7]: Did a line by line run through, and it matches with the previous `gitpod-installer-job.yaml`

File: install/installer/scripts/kots-install.sh
Comment (Pothulapati) [inline install/installer/scripts/kots-install.sh:30]: nit: Maybe document somewhere in the file about GITPOD_OBJECTS? :thinking: While its straightforward with other env's as they are input it could be confusing on what `kots-install.sh` emits.

**Review by Pothulapati — COMMENTED** [review]
Actually liking this approach, as it makes the kots -> install config a lot more simpler! The code looks good, will do a another pass with manual testing after which it should be ✅ 

Great Work! :)

File: install/installer/scripts/kots-install.sh
Comment (MrSimonEmms) [reply install/installer/scripts/kots-install.sh:74]: That's exactly the next step - good thought

File: install/installer/scripts/kots-install.sh
Comment (MrSimonEmms) [reply install/installer/scripts/kots-install.sh:30]: If it's ok with you, I won't fix in here - this won't be sticking around long as I've already started the next refactor and this probably won't survive. If it does, I'll fix it then

Comment (Pothulapati) [reply install/installer/scripts/kots-install.sh:30]: Sounds good!

File: install/installer/scripts/kots-install.sh
Comment (MrSimonEmms) [reply install/installer/scripts/kots-install.sh]: I've renamed it to `REG_DOCKER_CONFIG_ENABLED` to that keeps consistency with what I've done elsewhere (take the KOTS config name and captialise it). I've not changed the underlying KOTS config name (it's missing a `d`) in the expectation that someone could have started using it

**Review by Pothulapati — APPROVED** [review]
✅ from my side. Upgraded my Gitpod instance on AWS to use this version, and all the installer, status and preflight checks worked as expected.

**Pull request merged**
